src/main/java/com/iota/iri/utils/dag/DAGUtils.java

```diff
@@ -0,0 +1,366 @@
+     * @throws TraversalException if anything goes wrong while traversing the graph and processing the transactions
+     */
+    public void traverseApprovers(MilestoneViewModel startingMilestone,
+                                  TraversalCondition condition,
+                                  TraversalConsumer currentTransactionConsumer) throws TraversalException {
+        traverseApprovers(startingMilestone.getHash(), condition, currentTransactionConsumer, new HashSet<>());
+    }
+
+    //endregion ////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+    //region TRAVERSE APPROVEES (TOP -> BOTTOM) ////////////////////////////////////////////////////////////////////////
+
+    /**
+     * This method provides a generic interface for traversing all approvees of a transaction down to a given point.
+     *
+     * To make the use of this method as easy as possible we provide a broad range of possible parameter signatures, so
+     * we do not have to manually convert between milestones, transactions and hashes (see other methods with the same
+     * name).
+     *
+     * It uses an non-recursive iterative algorithm that is able to handle huge chunks of the tangle without running out
+     * of memory. It creates a queue of transactions that are being examined and processes them one by one. As new
+     * approvees are found, they will be added to the queue and processed accordingly.
+     *
+     * Every found transaction is passed into the provided condition lambda, to determine if it still belongs to the
+     * desired set of transactions and only then will be passed on to the currentTransactionConsumer lambda.
+     *
+     * @param startingTransaction the starting point of the traversal (does not get passed to the consumer)
+     * @param condition a lambda function that allows to control how long the traversal should continue
+     * @param currentTransactionConsumer a lambda function that allows us to "process" the found transactions
+     * @param processedTransactions a set of hashes that shall be considered as "processed" already and that will
+     *                              consequently be ignored in the traversal
+     * @throws TraversalException if anything goes wrong while traversing the graph and processing the transactions
+     */
+    public void traverseApprovees(TransactionViewModel startingTransaction,
+                                  TraversalCondition condition,
+                                  TraversalConsumer currentTransactionConsumer,
+                                  Set<Hash> processedTransactions) throws TraversalException {
+        final Queue<TransactionViewModel> transactionsToExamine = new LinkedList<>();
+        try {
+            transactionsToExamine.add(TransactionViewModel.fromHash(tangle, startingTransaction.getBranchTransactionHash()));
```

Comment:
duplicate lines later on, maybe it would siplify the code (in both implementations here) if the Queue is initialized w/ the `startingTransaction` and the traversal only happens in the `while`?